examples/pytorch/graphsage/link_pred.py

```diff
@@ -97,78 +80,91 @@ def inference(self, g, device, batch_size, num_workers, buffer_device=None):
             neg_dst = edge_split[split]['target_node_neg'].to(node_emb.device)
-            results.append(compute_mrr(model, node_emb, src, dst, neg_dst, device))
+            results.append(compute_mrr(model, evaluator, node_emb, src, dst, neg_dst, device))
     return results
 
-
-dataset = DglLinkPropPredDataset('ogbl-citation2')
-graph = dataset[0]
-graph, reverse_eids = to_bidirected_with_reverse_mapping(graph)
-graph = graph.to('cuda' if args.pure_gpu else 'cpu')
-reverse_eids = reverse_eids.to(device)
-seed_edges = torch.arange(graph.num_edges()).to(device)
-edge_split = dataset.get_edge_split()
-
-model = SAGE(graph.ndata['feat'].shape[1], 256).to(device)
-opt = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=5e-4)
-
-sampler = dgl.dataloading.NeighborSampler([15, 10, 5], prefetch_node_feats=['feat'])
-sampler = dgl.dataloading.as_edge_prediction_sampler(
+def train(args, device, g, reverse_eids, seed_edges, model):
+    # create sampler & dataloader
+    sampler = NeighborSampler([15, 10, 5], prefetch_node_feats=['feat'])
+    sampler = as_edge_prediction_sampler(
         sampler, exclude='reverse_id', reverse_eids=reverse_eids,
-        negative_sampler=dgl.dataloading.negative_sampler.Uniform(1))
-dataloader = dgl.dataloading.DataLoader(
-        graph, seed_edges, sampler,
+        negative_sampler=negative_sampler.Uniform(1))
+    use_uva = (args.mode == 'mixed')
+    dataloader = DataLoader(
+        g, seed_edges, sampler,
         device=device, batch_size=512, shuffle=True,
-        drop_last=False, num_workers=0, use_uva=not args.pure_gpu)
-
-durations = []
-for epoch in range(10):
-    model.train()
-    t0 = time.time()
-    for it, (input_nodes, pair_graph, neg_pair_graph, blocks) in enumerate(dataloader):
-        x = blocks[0].srcdata['feat']
-        pos_score, neg_score = model(pair_graph, neg_pair_graph, blocks, x)
-        pos_label = torch.ones_like(pos_score)
-        neg_label = torch.zeros_like(neg_score)
-        score = torch.cat([pos_score, neg_score])
-        labels = torch.cat([pos_label, neg_label])
-        loss = F.binary_cross_entropy_with_logits(score, labels)
-        opt.zero_grad()
-        loss.backward()
-        opt.step()
-        if (it + 1) % 20 == 0:
-            mem = torch.cuda.max_memory_allocated() / 1000000
-            print('Loss', loss.item(), 'GPU Mem', mem, 'MB')
-            if (it + 1) == 1000:
-                tt = time.time()
-                print(tt - t0)
-                durations.append(tt - t0)
-                break
-    if epoch % 10 == 0:
-        model.eval()
-        valid_mrr, test_mrr = evaluate(model, edge_split, device, 0 if args.pure_gpu else 12)
-        print('Validation MRR:', valid_mrr.item(), 'Test MRR:', test_mrr.item())
-print(np.mean(durations[4:]), np.std(durations[4:]))
+        drop_last=False, num_workers=0, use_uva=use_uva)
+    opt = torch.optim.Adam(model.parameters(), lr=0.0005)
```

Comment:
Previously, it was `lr=0.001, weight_decay=5e-4`; but then it always leads to a very bad testing MRR result (~0.66). Referring to ogb leaderboard [neighbor sampler implementation](https://github.com/snap-stanford/ogb/blob/master/examples/linkproppred/citation2/sampler.py), I adjusted to `lr=5e-4`, and it can result in a very close result to ogb leaderboard after 150 epochs (~0.805). With the default 10 epochs, it's around ~0.74 (see updated README).